### PR TITLE
Fix issue about not publishing data node updates

### LINKED ISSUE #2319: [🐛 BUG] Never receiving Data Node notifications/update with Core event
### What went wrong? 🤔

Taipy doesn't always publish the event of a data node update. Therefore, when a data node is changed, creating simple code to update my UI through Core events is impossible.

### Expected Behavior

I should receive all the appropriate events.

### Steps to Reproduce Issue

Run this code and the scenario:

```python
from taipy.core.notification import (
    CoreEventConsumerBase,
    Event,
    EventEntityType,
    EventOperation,
    Notifier,
)
import taipy as tp
import taipy.gui.builder as tgb
from taipy.gui import Gui, State, notify
from taipy.core import Scenario, DataNode
from taipy import Config

import time
import random
import pandas as pd


def some_task_1(input_data):
    time.sleep(2)
    print("Hello world")
    return "hello world"


def some_task_2(intermediate_data):
    time.sleep(2)
    print(f"{intermediate_data} from Taipy!")
    random_value = random.randint(0, 100)
    return f"{intermediate_data} from Taipy! {random_value}"


Config.configure_job_executions(mode="standalone", max_nb_of_workers=2)

input_data_cfg = Config.configure_csv_data_node(id="input_data", default_data={"x": [1, 2, 3], "y": [4, 5, 6]})
intermediate_data_cfg = Config.configure_data_node(id="intermediate_data")
output_data_cfg = Config.configure_data_node(id="output_data")

task_1_cfg = Config.configure_task(
    id="task_1",
    function=some_task_1,
    input=input_data_cfg,
    output=intermediate_data_cfg,
    skippable=True,
)

task_2_cfg = Config.configure_task(
    id="task_2",
    function=some_task_2,
    input=intermediate_data_cfg,
    output=output_data_cfg,
    skippable=True,
)

scenario_cfg = Config.configure_scenario(
    id="scenario",
    task_configs=[task_1_cfg, task_2_cfg],
)


def notify_user(state, data_node, scenarios):
    if state.scenario.id in scenarios:
        if data_node.config_id == "input_data":
            state.input_data = data_node.read()
        if data_node.config_id == "intermediate_data":
            notify(state, "success", "Task 1 is finished!")
        if data_node.config_id == "output_data":
            state.output_data = data_node.read()
            notify(state, "success", "Submission is finished!")


class SpecificCoreConsumer(CoreEventConsumerBase):
    def __init__(self, gui: Gui):
        self.gui = gui

        reg_id, queue = Notifier.register(
            # attribute_name="last_edit_date",
        )  # Adapt the registration to the events you want to listen to
        super().__init__(reg_id, queue)

    def process_event(self, event: Event):
        if (event.entity_type == EventEntityType.DATA_NODE
                and event.operation == EventOperation.UPDATE
                and event.attribute_name == "last_edit_date"):
            print(
                f"Received event: {event.operation} - {event.entity_type} - {event.entity_id}\n"
                f"                {event.attribute_name} - {event.attribute_value}\n"
            )
            data_node: DataNode = tp.get(event.entity_id)
            scenarios = [s.id for s in data_node.get_parents().get("scenario")]
            self.gui.broadcast_callback(
                notify_user,
                [data_node, scenarios],
            )

def change_scenario(state: State):
    state.input_data = state.scenario.input_data.read()
    state.output_data = state.scenario.output_data.read()

def run_scenario(state: State):
    print("Running scenario...")
    state.scenario.submit()
    print("Done!")

if __name__ == "__main__":
    tp.Orchestrator().run()
    scenario: Scenario = tp.create_scenario(scenario_cfg)
    input_data: pd.DataFrame = scenario.input_data.read()
    output_data: int = scenario.output_data.read()

    with tgb.Page() as page:
        tgb.scenario_selector(scenario, on_change=change_scenario)
        tgb.button("Run Scenario", on_action=run_scenario)
        tgb.chart("{input_data}")
        tgb.input("{output_data}")

    gui = Gui(page)
    SpecificCoreConsumer(gui).start()
    gui.run(debug=True)
```

### Versions of Taipy

- develop
- 4.0.1

### Acceptance Criteria

- [ ] A unit test reproducing the bug is added.
- [ ] Any new code is covered by a unit tested.
- [ ] Check code coverage is at least 90%.
- [ ] The bug reporter validated the fix.
- [ ] Related issue(s) in taipy-doc are created for documentation and Release Notes are updated.

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

### PR DESCRIPTION
**Pull request by jrobinAV**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Data node update events were not properly published when the update came from a task execution in standalone mode.

When a data node is written, 4 update events should be published:
- last_edit_date attribute update
- editor_id attribute update
- edit_in_progress attribute update
- editor_expiration_date attribute update
When the writing was made by a task execution, the events were published into a Notifier specific to the process, while it MUST be published to the main application Notifier.

## Related Tickets & Documents

- Closes #2319

## How to reproduce the issue

```python
from taipy.core import Status
from taipy.core.notification import (
    CoreEventConsumerBase,
    Event,
    EventEntityType,
    EventOperation,
    Notifier,
)
import taipy as tp
import taipy.gui.builder as tgb
from taipy.gui import Gui, State, notify
from taipy.core import Scenario, DataNode
from taipy import Config

import time
import pandas as pd


def some_task_1(input_data):
    time.sleep(2)
    return "hello world"


def some_task_2(intermediate_data):
    time.sleep(2)
    return f"{intermediate_data} from Taipy!"


Config.configure_job_executions(mode="standalone", max_nb_of_workers=2)

input_data_cfg = Config.configure_data_node(
    id="input_data", storage_type="csv", default_data={"x": [1, 2, 3], "y": [4, 5, 6]}
)
intermediate_data_cfg = Config.configure_data_node(
    id="intermediate_data",
)
output_data_cfg = Config.configure_data_node(
    id="output_data",
)

task_1_cfg = Config.configure_task(
    id="task_1",
    function=some_task_1,
    input=input_data_cfg,
    output=intermediate_data_cfg,
    skippable=True,
)

task_2_cfg = Config.configure_task(
    id="task_2",
    function=some_task_2,
    input=intermediate_data_cfg,
    output=output_data_cfg,
    skippable=True,
)

scenario_cfg = Config.configure_scenario(
    id="scenario",
    task_configs=[task_1_cfg, task_2_cfg],
)


def notify_user(state, data_node, scenarios):
    if [state.scenario.id](http://state.scenario.id/) in scenarios:
        if data_node.config_id == "input_data":
            state.input_data = data_node.read()
        if data_node.config_id == "intermediate_data":
            notify(state, "success", "Task 1 is finished!")
        if data_node.config_id == "output_data":
            state.output_data = data_node.read()
            notify(state, "success", "Submission is finished!")


class SpecificCoreConsumer(CoreEventConsumerBase):
    def __init__(self, gui: Gui):
        self.gui = gui

        reg_id, queue = Notifier.register(
            # attribute_name="last_edit_date",
        )  # Adapt the registration to the events you want to listen to
        super().__init__(reg_id, queue)

    def process_event(self, event: Event):
        if (
            event.entity_type == EventEntityType.DATA_NODE
            and event.operation == EventOperation.UPDATE
            and event.attribute_name == "last_edit_date"):
        ):
            print(
                f"Received event: {event.operation} - {event.entity_type} - {event.entity_id}\n"
                f"                {event.attribute_name} - {event.attribute_value}\n"
            )
            data_node: DataNode = tp.get(event.entity_id)
            scenarios = data_node.get_parents().get("scenario")
            self.gui.broadcast_callback(
                notify_user,
                [data_node, scenarios],
            )


def change_scenario(state: State):
    state.input_data = state.scenario.input_data.read()
    state.output_data = state.scenario.output_data.read()


def run_scenario(state: State):
    print("Running scenario...")
    state.scenario.submit()
    print("Done!")


def on_exception(state, function_name: str, ex: Exception):
    print(ex, "in ", function_name)


if __name__ == "__main__":
    tp.Orchestrator().run()
    scenario: Scenario = tp.create_scenario(scenario_cfg)

    input_data: pd.DataFrame = scenario.input_data.read()
    output_data: int = scenario.output_data.read()

    with [tgb.Page](http://tgb.page/)() as page:
        tgb.scenario_selector(scenario, on_change=change_scenario)
        tgb.button("Run Scenario", on_action=run_scenario)
        tgb.chart("{input_data}")
        tgb.input("{output_data}")

    gui = Gui(page)
    SpecificCoreConsumer(gui).start()
    gui.run(debug=True)
```

## branches or releases

- develop
- release/4.0

## Check list
- [x] Does this solution meet the acceptance criteria of the related issue?
- [x] Is the related issue checklist completed?
- [x] Does this PR add unit tests for the developed code?
- [x] End-to-End tests have been added or updated?
- [x] Have the release notes been updated?

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -32,9 +32,10 @@ _Describe which projects this change will impact and that needs to be backported
...
-- [ ] End-to-End tests have been added or updated?
-- [ ] Was the documentation updated, or a dedicated issue for documentation created? (If applicable)
-- [ ] Is the release notes updated? (If applicable)
+- [ ] This solution meet the acceptance criteria of the related issue.
+- [ ] The related issue checklist is completed.
+- [ ] This PR add unit tests for the developed code. If not, why?
+- [ ] End-to-End tests have been added or updated. If not, why?
+- [ ] The documentation has been updated, or a dedicated issue created. (If applicable)
+- [ ] The release notes have been updated? (If applicable)
+
```
Comment (jrobinAV) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
```

File: taipy/core/_orchestrator/_dispatcher/_task_function_wrapper.py [hunk 2]
```diff
@@ -65,7 +65,7 @@ def _write_data(self, outputs: List[DataNode], results, job_id: JobId):
                 for res, dn in zip(_results, outputs):
                     try:
                         data_node = data_manager._get(dn.id)
-                        data_node.write(res, job_id=job_id)
+                        data_node._write(res)
```
Comment (trgiangdo) [inline taipy/core/_orchestrator/_dispatcher/_task_function_wrapper.py:68]: Compare to the `DataNode.write()` method, seems like the `_DataManagerFactory._build_manager()._set()` line is missing.

Is it intended?

File: tests/core/notification/test_events_published.py [hunk 3]
```diff
@@ -156,6 +157,14 @@ def test_events_published_for_writing_dn():
 
 
 def test_events_published_for_scenario_submission():
+    _events_published_for_scenario_submission()
+
+def test_events_published_for_scenario_submission_standalone():
+    _events_published_for_scenario_submission(standalone=True)
+
+def _events_published_for_scenario_submission(standalone=False):
```
Comment (trgiangdo) [inline tests/core/notification/test_events_published.py]: Using the `@pytest.mark.parametrize(mode, ["development", "standalone"])` here would be cleaner.

However, this seems ok as well.

File: tests/core/notification/test_events_published.py [hunk 4]
```diff
@@ -176,7 +185,11 @@ def test_events_published_for_scenario_submission():
     # 1 submission update event for jobs
     # 3 submission update events (for status: PENDING, RUNNING and COMPLETED)
     # 1 submission update event for is_completed
-    scenario.submit()
+    if standalone:
+        Orchestrator().run()
```
Comment (trgiangdo) [inline tests/core/notification/test_events_published.py:185]: why do we only start the `Orchestrator()` in standalone mode and not development mode?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 5]
```diff
@@ -32,9 +32,9 @@ _Describe which projects this change will impact and that needs to be backported
...
 
-- [ ] Does this solution meet the acceptance criteria of the related issue?
-- [ ] Is the related issue checklist completed?
-- [ ] Does this PR adds unit tests for the developed code? If not, why?
-- [ ] End-to-End tests have been added or updated?
-- [ ] Was the documentation updated, or a dedicated issue for documentation created? (If applicable)
-- [ ] Is the release notes updated? (If applicable)
+- [ ] This solution meet the acceptance criteria of the related issue.
+- [ ] The related issue checklist is completed.
+- [ ] This PR add unit tests for the developed code. If not, why?
```
Comment (trgiangdo) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] This solution meets the acceptance criteria of the related issue.
- [ ] The related issue checklist is completed.
- [ ] This PR adds unit tests for the developed code. If not, why?
```

**Review by joaoandre-avaiga — DISMISSED** [review]

**Review by toan-quach — DISMISSED** [review]
lgtm!

File: taipy/core/_orchestrator/_dispatcher/_task_function_wrapper.py (hunk 2 above)
Comment (jrobinAV) [reply taipy/core/_orchestrator/_dispatcher/_task_function_wrapper.py:68]: Hum...
I think setting the data node here is useless. We don't update any attribute.

File: tests/core/notification/test_events_published.py (hunk 3 above)
Comment (jrobinAV) [reply tests/core/notification/test_events_published.py]: OK let me change that.

File: tests/core/notification/test_events_published.py (hunk 4 above)
Comment (jrobinAV) [reply tests/core/notification/test_events_published.py:185]: The development dispatcher does not need to be started. It is only a feature of the standalone dispatcher.

**Review by toan-quach — APPROVED** [review]

**Review by trgiangdo — APPROVED** [review]
LGTM

**Pull request merged**